### ISSUE TITLE
RethinkDB 2.3.5 server crashes on Ubuntu 14.04

### ISSUE THREAD

**Issue by RomanKisilenko** [rethinkdb/rethinkdb#6309]
Version: rethinkdb 2.3.5~0trusty (GCC 4.8.2)

```
2017-03-23T09:03:28.498254355 93891.344841s error: Error in src/rdb_protocol/changefeed.cc at line 1145:
2017-03-23T09:03:28.498288478 93891.344875s error: Guarantee failed: [inserted] 
2017-03-23T09:03:28.498297181 93891.344884s error: Backtrace:
2017-03-23T09:03:28.775297586 93891.621886s error: Thu Mar 23 09:03:28 2017\n\n1 [0xabda80]: backtrace_t::backtrace_t() at ??:?\n2 [0xabde13]: format_backtrace(bool) at ??:?\n3 [0xd
41b8d]: report_fatal_error(char const*, int, char const*, ...) at ??:?\n4 [0x8e4c28]: ql::changefeed::limit_manager_t::commit(rwlock_in_line_t*, boost::variant<ql::changefeed::prima
ry_ref_t, ql::changefeed::sindex_ref_t, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::va
riant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detai
l::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::
detail::variant::void_, boost::detail::variant::void_> const&) at ??:?\n5 [0x80f7c5]: /usr/bin/rethinkdb() [0x80f7c5] at 0x80f7c5 ()\n6 [0x8e1cfd]: ql::changefeed::server_t::foreach
_limit(boost::optional<std::string> const&, store_key_t const*, std::function<void (rwlock_in_line_t*, rwlock_in_line_t*, rwlock_in_line_t*, ql::changefeed::limit_manager_t*)>, auto
_drainer_t::lock_t const&) at ??:?\n7 [0x810c82]: rdb_update_single_sindex(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, un
signed long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::da
tum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*) at ??:?\n8 [0x818246]: callable_action_instance_t<std::_Bind<void (*(store_t*, store_t::sindex_access_t
*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std:
:pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*))(store_t*, store_t::sindex_access_t c
onst*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<
std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*)> >::run_action() at ??:?\n9 [0x9d
cb37]: coro_t::run() at ??:?
2017-03-23T09:03:28.775405500 93891.621992s error: Exiting.
```

**Comment by RomanKisilenko** [comment]
Hello, is there any way to work around this issue? We have this problem repeated regularly for us.

**Comment by srh** [comment]
I think this was fixed in the v2.3.x branch -- see https://github.com/rethinkdb/rethinkdb/issues/6093.

**Comment by RomanKisilenko** [comment]
Thank you for information. OK, we will wait until 2.3.6 release then.

**Comment by RomanKisilenko** [comment]
We've tested the latest code from 2.3.x branch and it fails with the same error message, but slightly different trace:
```
Error in src/rdb_protocol/changefeed.cc at line 1164:
Guarantee failed: [inserted]
Backtrace:
[0xce55b0]: backtrace_t::backtrace_t() at backtrace.cc:221 (discriminator 1)\n2 [0xce5943]: format_backtrace(bool) at backtrace.cc:330\n3 [0x7a23fd]: report_fatal_error(char const*, int, char const*, ...) at errors.cc:85\n4 [0x91ecf8]: ql::changefeed::limit_manager_t::commit(rwlock_in_line_t*, boost::variant<ql::changefeed::primary_ref_t, ql::changefeed::sindex_ref_t, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) at changefeed.cc:1164 (discriminator 1)\n5 [0x7f92f5]: /usr/local/bin/rethinkdb() [0x7f92f5] at 0x7f92f5 ()\n6 [0x91bb8c]: ql::changefeed::server_t::foreach_limit(boost::optional<std::string> const&, boost::optional<uuid_u> const&, store_key_t const*, std::function<void (rwlock_in_line_t*, rwlock_in_line_t*, rwlock_in_line_t*, ql::changefeed::limit_manager_t*)>, auto_drainer_t::lock_t const&) at optional.hpp:442\n7 [0x7fa810]: rdb_update_single_sindex(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*) at optional.hpp:438\n8 [0x8020f6]: callable_action_instance_t<std::_Bind<void (*(store_t*, store_t::sindex_access_t*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*))(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*)> >::run_action() at functional:1296\n9 [0x7f00f7]: coro_t::run() at coroutines.cc:282```

**Comment by RomanKisilenko** [comment]
Is there anything we can try to resolve the issue?

**Comment by AtnNn** [comment]
@RomanKisilenko What commit did you build from in v2.3.x? Can you isolate the crash to a specific query or action?

The backtrace suggests that this crash is triggered by a changefeed on a limit, something like `.limit(n).changes()`. You could avoid such queries until this bug is fixed.

**Comment by analytik** [comment]
This happened to me in the official 2.3.6 Debian build, three times in one evening.

```
Running on Linux 4.9.16-coreos-r1 x86_64
Version: rethinkdb 2.3.6~0jessie (GCC 4.9.2)
error: Error in src/rdb_protocol/changefeed.cc at line 1164:
error: Guarantee failed: [inserted]
error: Backtrace:
error: Tue Sep  5 16:59:39 2017

       1 [0x7d6e20]: backtrace_t::backtrace_t() at 0x7d6e20 (/usr/bin/rethinkdb)
       2 [0x7d7199]: format_backtrace(bool) at 0x7d7199 (/usr/bin/rethinkdb)
       3 [0x72b053]: report_fatal_error(char const*, int, char const*, ...) at 0x72b053 (/usr/bin/rethinkdb)
       4 [0x9cf735]: ql::changefeed::limit_manager_t::commit(rwlock_in_line_t*, boost::variant<ql::changefeed::primary_ref_t, ql::changefeed::sindex_ref_t, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) at 0x9cf735 (/usr/bin/rethinkdb)
       5 [0x99d482]: /usr/bin/rethinkdb() [0x99d482] at 0x99d482 ()
       6 [0x9ca95c]: ql::changefeed::server_t::foreach_limit(boost::optional<std::string> const&, boost::optional<uuid_u> const&, store_key_t const*, std::function<void (rwlock_in_line_t*, rwlock_in_line_t*, rwlock_in_line_t*, ql::changefeed::limit_manager_t*)>, auto_drainer_t::lock_t const&) at 0x9ca95c (/usr/bin/rethinkdb)
       7 [0x99ec17]: rdb_update_single_sindex(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*) at 0x99ec17 (/usr/bin/rethinkdb)
       8 [0x9a7a09]: callable_action_instance_t<std::_Bind<void (*(store_t*, store_t::sindex_access_t*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*))(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*)> >::run_action() at 0x9a7a09 (/usr/bin/rethinkdb)
       9 [0x815857]: coro_t::run() at 0x815857 (/usr/bin/rethinkdb)
error: Exiting.
```

**Comment by marshall007** [comment]
@analytik are you able to produce an example that does this reliably?

**Comment by analytik** [comment]
@marshall007 - Not really. I have a service that polls the `jobs` table every second or so and logs them, so I have three queries that are executed very often, but 1. there's no guarantee it's all of them, 2. I don't think even if I managed to create an anonymized data set, it would be reproducible without actually creating a load on the cluster.

Both with this, and other other bug reported (another guarantee failed, #6333), they happen sporadically - sometimes things work for a month, sometimes we get 3 crashes in a weekend. Both of them correlate with heavier load - we had an ORM doing sequential scan on a table with 2M rows every 5-10 seconds during the bad weekend, with version 2.3.5.

After upgrade to 2.3.6 it was smooth for two weeks (with the 2M sequential scan query now removed), but then the above happened 3x in one evening, on the same node, with CPU / network / HDD load seemingly normal. (E.g. 4 core / 8 HT core machines with about 60% CPU used for RethinkDB, ~130% used total on each machine; disk writes on the machine that crashed maybe 1~3MiB/s; network out 10MiB/s, network in 7MiB/s.)

Anyway, here are the queries:

```
r.table("foo")
 .orderBy({"index": r.desc("timestamp")})
 .limit(1)
 .changes({"includeInitial": true})
```

```
r.table("bar")
 .orderBy({"index": r.desc("created")})
 .limit(1)
 .changes({"squash": 0.250000, "includeInitial": true})
```

```
r.table("baz")
 .between(-0.800000, -0.050000,
   {"index": "someIndex", "leftBound": "open"})
 .orderBy({"index": "someIndex"})
 .filter(function (var19187) {
             return r.and(r.gt(r.row("someField"), r.minval), r.le(r.row("someField"), -100));
         })
 .limit(350)
 .changes({"squash": 0.250000, "includeInitial": true})
```

The first 2 could be refactored as a workaround - one of them is a heartbeat, the other just needs one latest value. The third one would be possibly a bit harder to rewrite, but if it meant more stable cluster, it would be worth it.

There's about 500 changefeeds going on in the cluster (most of them simple changefeeds on a single document), over 150 document writes per second, and some occasional reporting queries, so reproducing it might be hard.

Are there any debug builds of rethinkdb that would output more information on what are the possible causes? Something like "cannot do a thing with changefeed X because index Y was being updated with document Z".

**Comment by stellanhaglund** [comment]
I don't know if this helps but I got this to.

```
Version: rethinkdb 2.3.6~0trusty (GCC 4.8.2)
error: Error in src/rdb_protocol/changefeed.cc at line 1164:
error: Guarantee failed: [inserted] 
error: Backtrace:
addr2line: 'rethinkdb': No such file
addr2line: 'rethinkdb': No such file
error: Wed Nov 29 15:03:05 2017
       
       1 [0x808010]: backtrace_t::backtrace_t() at 0x808010 (rethinkdb)
       2 [0x8083a3]: format_backtrace(bool) at 0x8083a3 (rethinkdb)
       3 [0x7b9b2d]: report_fatal_error(char const*, int, char const*, ...) at 0x7b9b2d (rethinkdb)
       4 [0xc792e8]: ql::changefeed::limit_manager_t::commit(rwlock_in_line_t*, boost::variant<ql::changefeed::primary_ref_t, ql::changefeed::sindex_ref_t, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) at 0xc792e8 (rethinkdb)
       5 [0xb03965]: rethinkdb() [0xb03965] at 0xb03965 ()
       6 [0xc7617c]: ql::changefeed::server_t::foreach_limit(boost::optional<std::string> const&, boost::optional<uuid_u> const&, store_key_t const*, std::function<void (rwlock_in_line_t*, rwlock_in_line_t*, rwlock_in_line_t*, ql::changefeed::limit_manager_t*)>, auto_drainer_t::lock_t const&) at 0xc7617c (rethinkdb)
       7 [0xb04e80]: rdb_update_single_sindex(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*) at 0xb04e80 (rethinkdb)
       8 [0xb0c3d6]: callable_action_instance_t<std::_Bind<void (*(store_t*, store_t::sindex_access_t*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*))(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*)> >::run_action() at 0xb0c3d6 (rethinkdb)
       9 [0x8016e7]: coro_t::run() at 0x8016e7 (rethinkdb)
error: Exiting.
```

**Comment by srh** [comment]
For future reference the failing assertion is at

```
            bool inserted = item_queue.insert(pair).second;
            // We can never get two additions for the same key without a deletion
            // in-between.
            guarantee(inserted);
```

Apparently we can get two additions without a deletion in between.

**Comment by stellanhaglund** [comment]
is this only a problem in ubunut?

**Comment by srh** [comment]
This should be fixed in v2.4.x and next by https://github.com/rethinkdb/rethinkdb/pull/6710
In any case, a dupe of #6694.
